packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx

```diff
@@ -0,0 +1,228 @@
+    <FunctionResponse {...props} metadata={{ isRedacted: true, chunks }}>
+      ... The response is too big and hence redacted. The response can be queried using semantic similarity search by
+      calling the `loadBySimilarity` function.
+    </FunctionResponse>
+  );
+  // TODO: return a chunk based on query (gotta get it from conversation)
+}
+
+/**
+ * This function allows wrapping {@link FunctionResponse} elements that can possibly be too large.
+ * It will replace FunctionResponse elements with {@link LargeFunctionResponseHandler}s that know how to handle large responses.
+ *
+ * {@link LargeFunctionResponseHandler} will not modify responses that are not large. If they are, it will first try
+ * to optimize the response by dumping it as YAML, which is more token-efficient than JSON.
+ * If that doesn't work, it will split the response into chunks and allows the LLM to query it using semantic similarity
+ * search by exposing a dynamic function called `loadBySimilarity(query)`.
+ *
+ * Note that failed responses are not optimized and will simply be truncated to `failedMaxLength` characters.
+ *
+ * @see {@link redactedFunctionTools} to see how to add the `loadBySimilarity` function.
+ *
+ * @example
+ * ```tsx
+ *    <LargeFunctionResponseWrapper maxLength={4000} failedMaxLength={1000} numChunks={4}>
+ *      <ExecuteFunction func={tool[name].func} name={name} args={args} />
+ *    </LargeFunctionResponseWrapper>
+ * ```
+ */
+export async function LargeFunctionResponseWrapper(
+  { children, ...props }: { children: AI.Node } & LargeFunctionResponseProps,
+  { render }: AI.ComponentContext
+) {
+  // We need to render the children to get the FunctionResponse elements
+  const elements = await render(children, { stop: (e) => e.tag == FunctionResponse });
+
+  // We expect elements to just contain a single FunctionResponse but we handle multiple just in case
+  return elements.map((element) =>
+    AI.isElement(element) && element.tag == FunctionResponse ? (
+      <LargeFunctionResponseHandler {...props} {...(element.props as FunctionResponseProps)} />
+    ) : (
+      element
+    )
+  );
+}
+
+function getLastRedactedFnResponseData(messages: ConversationMessage[]): RedactedFuncionResponseMetadata | undefined {
+  let lastFnResData = undefined;
+  for (let i = messages.length - 1; i >= 0; i--) {
+    const msg = messages[i];
+    if (
+      msg.type == 'functionResponse' &&
+      typeof msg.element.props.metadata !== 'undefined' &&
+      'isRedacted' in msg.element.props.metadata &&
+      Boolean(msg.element.props.metadata.isRedacted) &&
+      'chunks' in msg.element.props.metadata &&
+      Array.isArray(msg.element.props.metadata.chunks) &&
+      msg.element.props.metadata.chunks.every((chunk) => typeof chunk === 'string')
+    ) {
+      lastFnResData = msg.element.props.metadata as unknown as RedactedFuncionResponseMetadata;
+    }
+  }
+  if (lastFnResData === undefined) {
+    return undefined;
+  }
+  return lastFnResData;
```

Comment:
```suggestion
  return lastFnResData;
```

Maybe I'm missing something, but that first condition seems unnecessary.

Reply:
Huh you're right.